### PR TITLE
Improve stack overflow reporting

### PR DESCRIPTION
**Pull request by janvorli**
This change modifies the SIGSEGV handling to use an alternate stack so that
we can safely detect and report stack overflow even in case when we are
really out of stack. Before, we were able to detect stack overflow and report
it only when JIT inserted stack probes (for functions with frames larger than
4kB) and so there was still space on the stack to run the sigsegv handler.
It brings in some additional complexity, since we need to switch to the original
stack of the thread once we figure out the sigsegv is not a stack overflow and
if we return from the hardware exception handler, we need to switch back to the
alternate stack before returning from the sigsegv handler.
Also, the alternate stack is created per thread and so we need to correctly destroy
it when a thread terminates and also install it on foreign threads that enter PAL.
This also requires creating fake stack frames to enable the libunwind to walk
the stack from the exception handler to the sigsegv location.

### CODE REVIEW

**Comment by janvorli** [comment]
@adityamandaleeka can you take a look please?
cc: @jkotas 
FYI: @sergiy-k, @Petermarcu 

I have tested it on x64, x86 and arm. I haven't tested the arm64 yet, but I will do that before merging this change.

**Comment by janvorli** [comment]
Fixes #1504

File: src/pal/src/exception/signal.cpp [hunk 1]
```diff
@@ -108,6 +118,75 @@ int g_common_signal_handler_context_locvar_offset = 0;
...
+        if (st != 0)
+        {
+            return FALSE;
+        }
+
+        stack_t ss;
+        ss.ss_sp = (char*)altStack;
+        ss.ss_size = altStackSize;
+        ss.ss_flags = 0;
+        sigaltstack(&ss, NULL);
```
Comment (adityamandaleeka) [inline src/pal/src/exception/signal.cpp]: Ensure that this has returned 0?

File: src/pal/src/include/pal/context.h [hunk 2]
```diff
@@ -640,6 +640,21 @@ LPVOID GetNativeContextPC(const native_context_t *context);
 
 /*++
 Function :
+    GetNativeContextSP
+
+    Returns the stack poiner from the native context.
```
Comment (adityamandaleeka) [inline src/pal/src/include/pal/context.h]: typo: pointer

File: src/pal/src/thread/context.cpp [hunk 3]
```diff
@@ -613,6 +613,35 @@ LPVOID GetNativeContextPC(const native_context_t *context)
 
 /*++
 Function :
+    GetNativeContextSP
+
+    Returns the stack poiner from the native context.
```
Comment (adityamandaleeka) [inline src/pal/src/thread/context.cpp]: typo: pointer

File: src/pal/src/exception/signal.cpp [hunk 4]
```diff
@@ -289,10 +396,38 @@ static void sigsegv_handler(int code, siginfo_t *siginfo, void *context)
...
+
+        // If the failure address is at most one page above or below the stack pointer, 
+        // we have a stack overflow. 
+        if ((failureAddress - (sp - VIRTUAL_PAGE_SIZE)) < 2 * VIRTUAL_PAGE_SIZE)
+        {
+            (void)write(STDERR_FILENO, StackOverflowMessage, sizeof(StackOverflowMessage) - 1);
+            PROCAbort();
+        }
+
+        // Now that we know the SIGSEGV didn't happen due to the stack overflow, execute the common
```
Comment (adityamandaleeka) [inline src/pal/src/exception/signal.cpp]: Nit: 'due to the stack overflow' -> 'due to a stack overflow'

**Comment by janvorli** [comment]
There are some issues, I am investigating ...

**Comment by janvorli** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test please

**Comment by janvorli** [comment]
After experimenting with alternative ways of implementing the unwinder helper functions, I have decided to go back to the multiple fake functions (although macro generated to write the code just once). It turned out to be the cleanest way.

**Comment by janvorli** [comment]
I have now verified that it works on ARM64 Linux too. While doing that, I have found that the debugger was not able to unwind through CallDescrWorkerInternal and so I have fixed it as part of this change.
CC: @sdmaclea

File: src/pal/inc/unixasmmacrosarm64.inc [hunk 5]
```diff
@@ -77,6 +78,7 @@ C_FUNC(\Name\()_End):
         .cfi_rel_offset \reg2, 8
         .ifc \reg1, fp
         mov fp, sp
+        .cfi_def_cfa_register fp
```
Comment (sdmaclea) [inline src/pal/inc/unixasmmacrosarm64.inc]: @janvorli Thanks

**Comment by adityamandaleeka** [comment]
LGTM

**Pull request merged**